### PR TITLE
Integrate Test262

### PR DESCRIPTION
**Pull request by jugglinmike**
Introduce a GNU Make target for retrieving TC-39's Test262 suite and
validating parsing of the files it contains. Interpret each file as a
parser test in accordance with that project's `INTERPRETING.md`
document. Allow for the specification of allowed failures via a
"whitelist" file so that the test suite may help prevent regressions in
this project in situations where this project has known bugs. Initialize
the "whitelist" file with a listing of all tests that are currently
failing. Extend the continuous integration environment's configuration
to automatically run these tests.

| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | yes
| Fixed tickets     | gh-633
| License           | MIT

Hey @hzoo!

Here's my first attempt at integrating Test262. There's a lot of new code here,
so I'm prepared to make a log of changes according to review feedback.

One thing I would prefer to do is split the `run_test262_utils.js` file into
multiple files, possibly within a dedicated directory. There's not much
precedent for that within the `scripts` directory, so I thought I'd hold off on
that until I heard back from you.

I'd also like to point out that the code within the `scripts` directory is not
currently being linted by the npm "lint" command. I manually satisfied the
linter anyway, with one exception: trailing commas. (I'm developing in Node.js
current LTS release where trailing commas are not supported.) Is there anything
you would like to change about this?

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
Travis reported this error:
```
$ if [ "$JOB" = "test262-test" ]; then make test-test262; fi
node scripts/run_test262.js
{ Error: ENOENT: no such file or directory, stat '/home/travis/build/babel/babylon/build/test262/test'
    at Error (native)
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/home/travis/build/babel/babylon/build/test262/test' }
```
Is it intended?

File: scripts/run_test262.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      console.log("");
+      console.log("Details:");
+      console.log(badnewsDetails.join("\n").replace(/^/gm, "   "));
+    }
+
+    process.exitCode = summary.passed ? 0 : 1;
+  })
+  .catch(function(err) {
+    console.error(err);
+    process.statusCode = 1;
```
Comment (nicolo-ribaudo) [inline scripts/run_test262.js]: This should be `process.exitCode`?

File: scripts/test262_whitelist.txt [hunk 2]
```diff
@@ -0,0 +1,7972 @@
+# This file lists tests that are known to produce incorrect results when parsed
+# with Babyline:
```
Comment (nicolo-ribaudo) [inline scripts/test262_whitelist.txt]: nit: Babylon

**Comment by nicolo-ribaudo** [comment]
I see there are a lot of failing tests which babylon actually supports via plugins, for example `language/expressions/async-generator/args-trailing-comma-multiple.js`.
I think there should be a way to specify which plugins enable for a given file/folder.

File: scripts/run_test262.js (hunk 1 above)
Comment (jugglinmike) [inline scripts/run_test262.js]: Yes. This is critical--thank you for spotting it.

**Comment by jugglinmike** [comment]
> I think there should be a way to specify which plugins enable for a given
> file/folder. Maybe we should use the features flag?

I'd be open to try that, but at some point, I think we should consider
introducing a YAML parser. The ad-hoc regular expressions I've written seem
"close enough" for the current use case, but using RegExps to interpret dynamic
values like that is much more error prone. I've been trying to avoid adding any
dependencies for this patch, so I'm wondering how the project maintainers would
feel about a new "devDependency" on a YAML parser.

**Comment by jugglinmike** [comment]
Alternatively, we could unconditionally enable all plugins for all tests. This is certainly less precision, but since Test262 generally doesn't include any "future hostile" tests, it may be an acceptable compromise.

**Comment by hzoo** [comment]
We just did something like this for our printer tests (generator) https://github.com/babel/babel/pull/6018 with explicit flags. We could do unconditional for all tests though for now, if someone runs the stage 0 preset, it would be the same thing and none of the plugins should be incompatible.

File: scripts/run_test262_utils.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      );
+    })
+    .then(flatten);
+};
+
+exports.runTest = function(test) {
+  const sourceType = test.isModule ? "module" : "script";
+
+  try {
+    parse(test.content, { sourceType: sourceType });
```
Comment (hzoo) [inline scripts/run_test262_utils.js]: Just need to add the ones in https://github.com/babel/babylon#plugins

**Comment by hzoo** [comment]
is there already a script/way to remove the whitelist tests automatically if things are fixed later?

File: scripts/run_test262.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+const path = require("path");
+const chalk = require("chalk");
+const utils = require("./run_test262_utils");
+
+const testDir = path.join(__dirname, "..", "build", "test262", "test");
+const whitelistFile = path.join(__dirname, "test262_whitelist.txt");
+
+Promise.all([utils.getTests(testDir), utils.getWhitelist(whitelistFile)])
+  .then(function([tests, whitelist]) {
```
Comment (Kovensky) [inline scripts/run_test262.js]: This breaks in Node 4 (do we care?)

File: scripts/run_test262.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        " invalid programs did not produce a parsing error" +
+        " (and allowed by the whitelist file)",
+      summary.allowed.falseNegative.length +
+        " valid programs produced a parsing error" +
+        " (and allowed by the whitelist file)",
+    ];
+    const badnews = [];
+    const badnewsDetails = [];
+
+    [
```
Comment (Kovensky) [inline scripts/run_test262.js]: Lines starting with a `[` are pretty scary, regardless of semicolons.

Maybe save this into a `const`, or make it the right side of a `for of`?

File: scripts/run_test262.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      if (!tests.length) {
+        return;
+      }
+
+      const desc = tests.length + " " + label;
+
+      badnews.push(desc);
+      badnewsDetails.push(desc + ":");
+      badnewsDetails.push(
+        ...tests.map(function(test) {
```
Comment (Kovensky) [inline scripts/run_test262.js]: This also breaks in Node 4 😄

File: scripts/run_test262_utils.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+function hasEarlyError(src) {
+  return !!(
+    src.match(/^\s*negative:\s*$/m) && src.match(/^\s+phase:\s*early\s*$/m)
+  );
+}
+
+function readDir(dirName) {
+  return new Promise(function(resolve, reject) {
+    fs.readdir(dirName, function(err, contents) {
```
Comment (Kovensky) [inline scripts/run_test262_utils.js]: Any objections to adding a devDependency on [`util.promisify`](https://www.npmjs.com/package/util.promisify)?

It's available natively in Node >= 8, the package just shims it if it's missing.

File: scripts/run_test262_utils.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+  const sourceType = test.isModule ? "module" : "script";
+
+  try {
+    parse(test.content, { sourceType: sourceType });
+    test.actualError = false;
+  } catch (err) {
+    test.actualError = true;
+  }
+
+  test.result = test.expectedError ^ test.actualError ? "fail" : "pass";
```
Comment (Kovensky) [inline scripts/run_test262_utils.js]: Better to use `!==`.

File: scripts/run_test262_utils.js [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+exports.getWhitelist = function(filename) {
+  return readFile(filename).then(function(contents) {
+    return contents
+      .split("\n")
+      .map(function(line) {
+        return line.replace(/#.*$/, "").trim();
+      })
+      .filter(function(line) {
+        return line.length > 0;
```
Comment (Kovensky) [inline scripts/run_test262_utils.js]: `.filter(Boolean)` would work too but I guess this is clearer

File: scripts/run_test262.js (hunk 4 above)
Comment (hzoo) [inline scripts/run_test262.js]: I think it's ok to only run on latest for these?

File: scripts/run_test262.js (hunk 5 above)
Comment (jugglinmike) [inline scripts/run_test262.js]: For consistency, I'd prefer to keep a functional style. But I see your point. I've made this the right-hand side of a `void` expression--that will guard against ASI-related bugs, and it should also more clearly indicate to the reader that this literal value is not used beyond this statement.

File: scripts/run_test262_utils.js (hunk 7 above)
Comment (jugglinmike) [inline scripts/run_test262_utils.js]: I wouldn't object, but I believe this question was @hzoo, so I'm going to hold off unless he agrees.

**Comment by jugglinmike** [comment]
> is there already a script/way to remove the whitelist tests automatically if
> things are fixed later?

There was not, but that's a good idea. Maintaining the white list in JSHint has
been a pain point for me. I've updated the patch to include a command-line
flag and dedicated Makefile target.

So I think that's everything--this should be ready for another round of review.

File: scripts/run_test262_utils.js (hunk 7 above)
Comment (hzoo) [inline scripts/run_test262_utils.js]: I don't think we need to run on anything other than latest node (8)? We can add it too if we really want to but I think it's fine to run one one version of node like we do for coverage and lint

Comment (jugglinmike) [inline scripts/run_test262_utils.js]: Works for me!

**Comment by jugglinmike** [comment]
I've updated the job to run in the latest release of Node.js, and I've taken @Kovensky's advice regarding the "util.promisify" module. My thinking is that many contributors likely prefer to develop in the LTS release. Even though they may not all care to run this particular test suite, the module's dependency tree is light enough to be justify the convenience.

The `babel-test` job is failing. Is there something we should do about that in this patch?

**Comment by hzoo** [comment]
> There was not, but that's a good idea. 

Like it would basically run the whitelist tests and if they pass correctly instead of erroring/not erroring it could tell you to remove them (or it is done automatically), similar to updating a snapshot/fixture

**Comment by hzoo** [comment]
oops I just merged https://github.com/babel/babel/pull/6056, which means we need to update yarn at least?

**Comment by jugglinmike** [comment]
Whoops; before you updated the comment, I thought you were saying I should update the yarn lock file. I guess wasn't necessary, after all (I'm not familiar with that tool). Should I revert this latest commit?

**Comment by hzoo** [comment]
I updated Babel to use a new feature in yarn; I think we have to add https://github.com/babel/babel/pull/6056/files#diff-354f30a63fb0907d4ad57269548329e3L15 since the yarn version it's using now is older. Not sure it's going to fix but it's an issue due to us targeting master. Either way it would of broke in someone else's PR

**Comment by hzoo** [comment]
Yeah we don't need that commit, I can try this locally and push a change

**Comment by danez** [comment]
I just tested this also locally and it doesn't work and stops with `EMFILE: too many open files`. Seems reading all the tests async without limit is not a good idea, but works on travis for some reason.

**Comment by nicolo-ribaudo** [comment]
We could use [graceful-fs](https://github.com/isaacs/node-graceful-fs), which handles EMFILE errors automatically.

**Comment by hzoo** [comment]
Yeah graceful-fs sounds good

**Comment by hzoo** [comment]
Thanks for starting this @jugglinmike!

File: Makefile [hunk 10]
```diff
@@ -30,3 +31,14 @@ bootstrap-flow: clean
 
 test-flow:
 	node scripts/run_flow_tests.js
+
+bootstrap-test262: clean
+	mkdir ./build
+	git clone https://github.com/tc39/test262.git ./build/test262
```
Comment (hzoo) [inline Makefile]: could we do a depth=x here as well?

**Comment by hzoo** [comment]
screen shot 2017-08-06 at 9 25 41 pm

**Comment by jugglinmike** [comment]
It's been my pleasure :)

**Pull request merged**